### PR TITLE
KA-389 Add a GetCodename method to ICodeFirstTypeProvider

### PR DESCRIPTION
**Pull request by Neytus**
The PR fixes this [issue](https://github.com/Kentico/delivery-sdk-net/issues/101).
A `GetCodename` method signature is introduced to the `ICodeFirstTypeProvider` interface.

Internal ref: KA-389

### CODE REVIEW

File: KenticoCloud.Delivery.Tests/Models/ContentTypes/CustomTypeProvider.cs [hunk 1]
```diff
@@ -1,43 +1,46 @@
...
-                case "office":
-                    return typeof(Office);
-                case "tweet":
-                    return typeof(Tweet);
-                default:
-                    return null;
+                return null;
             }
+
+            return TypesDictionary.Keys.First(type => GetCodename(type).Equals(contentType));
```
Comment (petrsvihlik) [inline KenticoCloud.Delivery.Tests/Models/ContentTypes/CustomTypeProvider.cs]: wouldn't `FirstOrDefault` do the trick?

File: KenticoCloud.Delivery/CodeFirst/ICodeFirstTypeProvider.cs [hunk 2]
```diff
@@ -12,5 +12,11 @@ public interface ICodeFirstTypeProvider
         /// </summary>
         /// <param name="contentType">Content type identifier.</param>
         Type GetType(string contentType);
+
+        /// <summary>
+        /// Returns a codename corresponding to the given content type model.
+        /// </summary>
+        /// <param name="contentType">Content type model.</param>
+        string GetCodename(Type contentType);
```
Comment (petrsvihlik) [inline KenticoCloud.Delivery/CodeFirst/ICodeFirstTypeProvider.cs]: now you can rewrite or maybe even get rid of the `ContentTypeExtractor` class

Comment (petrsvihlik) [reply KenticoCloud.Delivery/CodeFirst/ICodeFirstTypeProvider.cs]: but make sure you keep the tests that check whether the `system.type` query param is added

File: KenticoCloud.Delivery/DeliveryClient.cs [hunk 3]
```diff
@@ -590,5 +589,30 @@ private async Task<JObject> GetResponseContent(HttpResponseMessage httpResponseM
...
+
+        internal IEnumerable<IQueryParameter> ExtractParameters<T>(IEnumerable<IQueryParameter> parameters = null)
+        {
+            var enhancedParameters = parameters != null
+                ? new List<IQueryParameter>(parameters)
+                : new List<IQueryParameter>();
+
+            var codename = _codeFirstModelProvider.TypeProvider.GetCodename(typeof(T));
+
+            if (!IsAlreadyInParameters(parameters) && codename != null)
```
Comment (vaclavm) [inline KenticoCloud.Delivery/DeliveryClient.cs]: Performance hint: Switch the check, if codename is null there is no need to check the params.

File: KenticoCloud.Delivery/DeliveryClient.cs [hunk 4]
```diff
@@ -590,5 +589,30 @@ private async Task<JObject> GetResponseContent(HttpResponseMessage httpResponseM
...
+            {
+                enhancedParameters.Add(new EqualsFilter("system.type", codename));
+            }
+            return enhancedParameters;
+        }
+
+        private static bool IsAlreadyInParameters(IEnumerable<IQueryParameter> parameters)
```
Comment (vaclavm) [inline KenticoCloud.Delivery/DeliveryClient.cs]: What is in the parameters? Please provide better name, something like IsTypeInParameters().

**Pull request merged**